File: iterableapi/src/main/java/com/iterable/iterableapi/InAppAnimationService.java
package com.iterable.iterableapi;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.TransitionDrawable;
import android.view.View;
import android.view.Window;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.ColorUtils;

/**
 * Service class for in-app message animations.
 * Centralizes animation logic shared between Fragment and Dialog implementations.
 */
class InAppAnimationService {

    private static final int ANIMATION_DURATION_MS = 300;
    private static final String TAG = "InAppAnimService";

    /**
     * Creates a background drawable with the specified color and alpha
     * @param hexColor The background color in hex format (e.g., "#000000")
     * @param alpha The alpha value (0.0 to 1.0)
     * @return ColorDrawable with the specified color and alpha, or null if parsing fails
     */
    @Nullable
    public ColorDrawable createInAppBackgroundDrawable(@Nullable String hexColor, double alpha) {
        int backgroundColor;
        
        try {
            if (hexColor != null && !hexColor.isEmpty()) {
                backgroundColor = Color.parseColor(hexColor);
            } else {
                backgroundColor = Color.BLACK;
            }
        } catch (IllegalArgumentException e) {
            IterableLogger.w(TAG, "Invalid background color: " + hexColor + ". Using BLACK.", e);
            backgroundColor = Color.BLACK;
        }

        int backgroundWithAlpha = ColorUtils.setAlphaComponent(
            backgroundColor,
            (int) (alpha * 255)
        );

        return new ColorDrawable(backgroundWithAlpha);
    }

    /**
     * Animates the window background from one drawable to another
     * @param window The window to animate
     * @param from The starting drawable
     * @param to The ending drawable
     * @param shouldAnimate If false, sets the background immediately without animation
     */
    public void animateWindowBackground(@NonNull Window window, @NonNull Drawable from, @NonNull Drawable to, boolean shouldAnimate) {
        if (shouldAnimate) {
            Drawable[] layers = new Drawable[]{from, to};
            TransitionDrawable transition = new TransitionDrawable(layers);
            window.setBackgroundDrawable(transition);
            transition.startTransition(ANIMATION_DURATION_MS);
        } else {
            window.setBackgroundDrawable(to);
        }
    }

    /**
     * Shows the in-app background with optional fade-in animation
     * @param window The window to set the background on
     * @param hexColor The background color in hex format
     * @param alpha The background alpha (0.0 to 1.0)
     * @param shouldAnimate Whether to animate the background fade-in
     */
    public void showInAppBackground(@NonNull Window window, @Nullable String hexColor, double alpha, boolean shouldAnimate) {
        ColorDrawable backgroundDrawable = createInAppBackgroundDrawable(hexColor, alpha);
        
        if (backgroundDrawable == null) {
            IterableLogger.w(TAG, "Failed to create background drawable");
            return;
        }

        if (shouldAnimate) {
            // Animate from transparent to the target background
            ColorDrawable transparentDrawable = new ColorDrawable(Color.TRANSPARENT);
            animateWindowBackground(window, transparentDrawable, backgroundDrawable, true);
        } else {
            window.setBackgroundDrawable(backgroundDrawable);
        }
    }

    /**
     * Shows and optionally animates a WebView
     * @param webView The WebView to show
     * @param shouldAnimate Whether to animate the appearance
     * @param context Context for loading animation resources (only needed if shouldAnimate is true)
     */
    public void showAndAnimateWebView(@NonNull View webView, boolean shouldAnimate, @Nullable Context context) {
        if (shouldAnimate && context != null) {
            // Animate with alpha fade-in
            webView.setAlpha(0f);
            webView.setVisibility(View.VISIBLE);
            webView.animate()
                .alpha(1.0f)
                .setDuration(ANIMATION_DURATION_MS)
                .start();
        } else {
            // Show immediately
            webView.setAlpha(1.0f);
            webView.setVisibility(View.VISIBLE);
        }
    }

    /**
     * Hides the in-app background with optional fade-out animation
     * @param window The window to modify
     * @param hexColor The current background color
     * @param alpha The current background alpha
     * @param shouldAnimate Whether to animate the background fade-out
     */
    public void hideInAppBackground(@NonNull Window window, @Nullable String hexColor, double alpha, boolean shouldAnimate) {
        if (shouldAnimate) {
            ColorDrawable backgroundDrawable = createInAppBackgroundDrawable(hexColor, alpha);
            ColorDrawable transparentDrawable = new ColorDrawable(Color.TRANSPARENT);
            
            if (backgroundDrawable != null) {
                animateWindowBackground(window, backgroundDrawable, transparentDrawable, true);
            }
        } else {
            window.setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        }
    }

    /**
     * Prepares a view to be shown by hiding it initially
     * This is typically called before the resize operation
     * @param view The view to hide
     */
    public void prepareViewForDisplay(@NonNull View view) {
        view.setAlpha(0f);
        view.setVisibility(View.INVISIBLE);
    }
}

File: iterableapi/src/main/java/com/iterable/iterableapi/InAppAnimationService.kt
package com.iterable.iterableapi

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.graphics.drawable.TransitionDrawable
import android.view.View
import android.view.Window
import android.view.animation.AnimationUtils
import androidx.annotation.AnimRes
import androidx.annotation.RestrictTo
import androidx.core.graphics.ColorUtils

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
internal class InAppAnimationService {

    fun createInAppBackgroundDrawable(hexColor: String?, alpha: Double): ColorDrawable? {
        val backgroundColor = try {
            if (!hexColor.isNullOrEmpty()) {
                Color.parseColor(hexColor)
            } else {
                Color.BLACK
            }
        } catch (e: IllegalArgumentException) {
            IterableLogger.w(TAG, "Invalid background color: $hexColor. Using BLACK.", e)
            Color.BLACK
        }

        val backgroundWithAlpha = ColorUtils.setAlphaComponent(
            backgroundColor,
            (alpha * 255).toInt()
        )

        return ColorDrawable(backgroundWithAlpha)
    }

    fun animateWindowBackground(window: Window, from: Drawable, to: Drawable, shouldAnimate: Boolean) {
        if (shouldAnimate) {
            val layers = arrayOf(from, to)
            val transition = TransitionDrawable(layers)
            window.setBackgroundDrawable(transition)
            transition.startTransition(IterableConstants.ITERABLE_IN_APP_BACKGROUND_ANIMATION_DURATION)
        } else {
            window.setBackgroundDrawable(to)
        }
    }

    fun showInAppBackground(window: Window, hexColor: String?, alpha: Double, shouldAnimate: Boolean) {
        val backgroundDrawable = createInAppBackgroundDrawable(hexColor, alpha)

        if (backgroundDrawable == null) {
            IterableLogger.w(TAG, "Failed to create background drawable")
            return
        }

        if (shouldAnimate) {
            val transparentDrawable = ColorDrawable(Color.TRANSPARENT)
            animateWindowBackground(window, transparentDrawable, backgroundDrawable, true)
        } else {
            window.setBackgroundDrawable(backgroundDrawable)
        }
    }

    /**
     * Returns the enter animation resource for the given in-app layout, mirroring the
     * behavior of [IterableInAppFragmentHTMLNotification] so Compose/Dialog hosts get the
     * same slide/fade animations as Fragment hosts.
     */
    @AnimRes
    fun getEnterAnimationResource(layout: InAppLayoutService.InAppLayout): Int {
        return when (layout) {
            InAppLayoutService.InAppLayout.TOP -> R.anim.slide_down_custom
            InAppLayoutService.InAppLayout.BOTTOM -> R.anim.slide_up_custom
            InAppLayoutService.InAppLayout.CENTER,
            InAppLayoutService.InAppLayout.FULLSCREEN -> R.anim.fade_in_custom
        }
    }

    /**
     * Returns the exit animation resource for the given in-app layout, mirroring the
     * behavior of [IterableInAppFragmentHTMLNotification].
     */
    @AnimRes
    fun getExitAnimationResource(layout: InAppLayoutService.InAppLayout): Int {
        return when (layout) {
            InAppLayoutService.InAppLayout.TOP -> R.anim.top_exit
            InAppLayoutService.InAppLayout.BOTTOM -> R.anim.bottom_exit
            InAppLayoutService.InAppLayout.CENTER,
            InAppLayoutService.InAppLayout.FULLSCREEN -> R.anim.fade_out_custom
        }
    }

    fun showAndAnimateWebView(
        webView: View,
        shouldAnimate: Boolean,
        context: Context?,
        layout: InAppLayoutService.InAppLayout
    ) {
        webView.alpha = 1.0f
        webView.visibility = View.VISIBLE

        if (shouldAnimate && context != null) {
            try {
                val anim = AnimationUtils.loadAnimation(context, getEnterAnimationResource(layout))
                anim.duration = IterableConstants.ITERABLE_IN_APP_ANIMATION_DURATION.toLong()
                webView.startAnimation(anim)
            } catch (e: Exception) {
                IterableLogger.w(TAG, "Failed to start enter animation", e)
            }
        }
    }

    /**
     * Plays the layout-appropriate exit animation on the given view. Returns `true` when
     * an animation was started, `false` otherwise (either because [shouldAnimate] was
     * false or loading the animation failed). Callers should schedule dismissal
     * accordingly.
     */
    fun hideAndAnimateWebView(
        webView: View,
        shouldAnimate: Boolean,
        context: Context?,
        layout: InAppLayoutService.InAppLayout
    ): Boolean {
        if (!shouldAnimate || context == null) {
            return false
        }
        return try {
            val anim = AnimationUtils.loadAnimation(context, getExitAnimationResource(layout))
            anim.duration = IterableConstants.ITERABLE_IN_APP_ANIMATION_DURATION.toLong()
            webView.startAnimation(anim)
            true
        } catch (e: Exception) {
            IterableLogger.w(TAG, "Failed to start exit animation", e)
            false
        }
    }

    fun hideInAppBackground(window: Window, hexColor: String?, alpha: Double, shouldAnimate: Boolean) {
        if (shouldAnimate) {
            val backgroundDrawable = createInAppBackgroundDrawable(hexColor, alpha)
            val transparentDrawable = ColorDrawable(Color.TRANSPARENT)

            if (backgroundDrawable != null) {
                animateWindowBackground(window, backgroundDrawable, transparentDrawable, true)
            }
        } else {
            window.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        }
    }

    fun prepareViewForDisplay(view: View) {
        view.alpha = 0f
        view.visibility = View.INVISIBLE
    }

    companion object {
        private const val TAG = "InAppAnimService"
    }
}

